### PR TITLE
Update docstrings and api documentation.

### PR DESCRIPTION
**Pull request by beadysea**
I've updated the docstrings for all of `bluetooth/bluez` and some others in the other bluetooth modules. Also set up sphinx autosummary templates to allow sphinx to generate stub files. All of the files under `docs/api/packages/` from modules down are autogenerated by sphinx, creating the file structure which readthedocs will build against. 

Tested with sphinx in nitpick mode with no errors so hopefully this should be painless at the readthedocs end.  

Key points for review:

~~- I commented out the `__doc__` sections in bluetooth __init__.py. Sphinx seems to be a bit adhoc in finding those references.  I think it has to do with imports, so if `from bluetooth import *` is in a file sphinx finds the __doc__, if a file just does `import Bluetooth` or `from bluetooth import btcommon` then __doc__ is not read.  Anyway I decided to just put the documentation in each separate file. Duplication I know but then I'd argue it saves a developer from hunting around the codebase looking for the documented spec.~~
- widcomm.py  I had to comment out line 107 to stop sphinx falling over. It looks to me that this line has been left in from some testing. If it needs to be there then I think we have to do: `if __name__ == __main__` etc to prevent sphinx trying to execute it.
- I chose to write the docstrings using numpy docstring style. It does have the downside that it takes up more lines, however it does produce a clean easy to read docstring in code and it renders it out very nicely for readthedocs.
- Some of the parameter or return types may be wrong. I took my best guess at them by digging around the codebase and trying to deduce their type. As mentioned I focussed on bluez first. If that is accurate then everything mostly becomes a copy over to the other modules.

Final point - I think this is the last huge commit. From now on it will just be adding or amending pages in docs, or docstring updates on the six source code files.

### CODE REVIEW

**Comment by rgov** [comment]
Thank you for all the work you've put into this.

One thing I'd like to think more about is the first bullet, commenting out the docstrings. If we do this, it will also break using the built-in `help()` function and probably code completion hints in smart IDEs.

What can we do to make this better? Most projects don't have this problem, do they?

**Comment by beadysea** [comment]
I'll try a local sphinx build with the init.py docstrings in again and see what happens.

**Comment by rgov** [comment]
Do you know if there is a way to get Read the Docs to run the sphinx-apidoc generation manually? Then we can drop these from the repo. I'm concerned that if we update the documentation, we might forget to run the generation step, and the code and docs will diverge.

**Comment by beadysea** [comment]
> Do you know if there is a way to get Read the Docs to run the sphinx-apidoc generation manually? Then we can drop these from the repo. I'm concerned that if we update the documentation, we might forget to run the generation step, and the code and docs will diverge.

I think my latest commit solves this. I pointed autosummary to build api pages in docs _build directory and set autosummary generate in conf.py. Tested locally with sphinx-build and html pages were generated with no errors.

**Comment by rgov** [comment]
I see, the point of the docstring override like this is to provide a single documentation for different backend implementations.

The line `inquirer = WCInquirer ()` in widcomm.py would not be appropriate to hide behind `__name__ == '__main__'` because it is a global shared instance of WCInquirer that is used throughout module's functions.  What is the Sphinx error? It's unusual, because it is a pretty common pattern.

**Comment by beadysea** [comment]
See sphinx document extract:
```
autodoc imports the modules to be documented. 
If any modules have side effects on import, these will be executed by autodoc when 
sphinx-build is run.  If you document scripts (as opposed to library modules), make 
sure their main routine is protected by a if __name__ == '__main__' condition.
```
With `inquirer = WCInquirer ()` in place, sphinx falls over because it tries to execute the statement.

https://try-rtd-l10n.readthedocs.io/ja/latest/usage/quickstart.html#autodoc

**Comment by rgov** [comment]
Executing the statement isn't precisely the problem, it's that the `__init__` method does something (tries to use a C module that might not be available and/or connects to a local port).

Potentially we should rewrite the code so that nothing happens at `__init__` time? I'm not sure. Is there a global variable or environment variable or something that tells us Autodoc is running?

**Comment by rgov** [comment]
Is any of the content here needed ? https://github.com/pybluez/pybluez/tree/wiki If not we can delete this branch.

**Comment by beadysea** [comment]
I have a local copy of the wiki branch so it's safe to delete. If there is anything missing in the current docs (which I doubt) then we can fix.

**Comment by beadysea** [comment]
I have decided to leave widcomm out of the auto generated docs until we figure out how to get sphinx to play nicely with the code in that module.

**Comment by rgov** [comment]
That sounds reasonable.

**Comment by ukBaz** [comment]
FYI: If you have module that is not available for import by sphinx then you can use mock:
http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_mock_imports

**Comment by beadysea** [comment]
Thanks @ukBaz, I've already applied mocks for _widcomm among others in conf.py

File: bluetooth/__init__.py [hunk 1]
```diff
@@ -48,106 +54,200 @@ def _dbg(*args):
...
-
-    if lookup_names is False, returns a list of bluetooth addresses.
-    if lookup_names is True, returns a list of (address, name) tuples
-
-    lookup_names=False
-    if set to True, then discover_devices also attempts to lookup the
-    display name of each detected device.
-
-    if lookup_class is True, the class of the device is added to the tuple
+    """Perform a bluetooth device discovery.
```
Comment (rgov) [inline bluetooth/__init__.py]: When referring to the Bluetooth protocol (as opposed to the `bluetooth` module), please capitalize it.

File: bluetooth/__init__.py [hunk 2]
```diff
@@ -48,106 +54,200 @@ def _dbg(*args):
...
+        
+    name : str
+        The name of the service and service_id (if specified). This should be a string
+        of the form "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX", where each 'X' is a hexadecimal
+        digit.
+
+    service_classes : list 
+        a list of service classes belonging to the advertised service.
+
+        Each service class is represented by a Universal Unique Identifier (UUID). This function
```
Comment (rgov) [inline bluetooth/__init__.py:69]: "Universally"; or just drop the part that spells it out, I think UUIDs are widely known.

File: bluetooth/__init__.py [hunk 3]
```diff
@@ -48,106 +54,200 @@ def _dbg(*args):
...
+    
+    Parameters
+    ----------
+    sock : str
+        The BluetoothSocket to stop advertising service on.
+    
+    Raises
+    ------
+    BluetoothError
+        When sdp fails to stop advertising for some reason.
```
Comment (rgov) [inline bluetooth/__init__.py]: Capitalize "SDP".

File: bluetooth/bluez.py [hunk 4]
```diff
@@ -1,20 +1,59 @@
-import array
-import fcntl
+"""GNU/Linux implementation of Bluetooth."""
```
Comment (rgov) [inline bluetooth/bluez.py:1]: This is not actually an implementation of "Bluetooth", perhaps you mean "of the bluetooth module"?

File: bluetooth/bluez.py [hunk 5]
```diff
@@ -1,20 +1,59 @@
...
 import struct
 from errno import (EADDRINUSE, EBUSY, EINVAL)
 
-from bluetooth.btcommon import *
-import bluetooth._bluetooth as _bt
-from bluetooth._bluetooth import HCI, RFCOMM, L2CAP, SCO, SOL_L2CAP, \
-                                    SOL_RFCOMM, L2CAP_OPTIONS
-
-get_byte = ord if sys.version_info.major < 3 else int
+if sys.version_info.major < 3:
```
Comment (rgov) [inline bluetooth/bluez.py:6]: In a recent pull request we were advised to check for features and not version numbers. Can we turn this into a feature check?

```
try:
    import x
except ImportError:
    import y
```

or

```
import x
try:
    a = x.old
except ValueError:
    a = x.new
```

File: bluetooth/bluez.py [hunk 6]
```diff
@@ -1,20 +1,59 @@
...
+    from .btcommon import *
+    import _bluetooth as _bt
+    get_byte = ord
+else:
+    from bluetooth.btcommon import *
+    import bluetooth._bluetooth as _bt
+    get_byte = int
+import array
```
Comment (rgov) [inline bluetooth/bluez.py:14]: Wasn't this just removed in another pull request?

File: bluetooth/bluez.py [hunk 7]
```diff
@@ -49,6 +88,20 @@ def discover_devices (duration=8, flush_cache=True, lookup_names=False,
...
+    
+    Returns
+    -------
+    list
+        A list of bluetooth address strings.
+
+    Raises
+    ------
+    BluetoothError
+        When a bluetooth hci error occurs.
```
Comment (rgov) [inline bluetooth/bluez.py:101]: Please capitalize HCI.

File: bluetooth/bluez.py [hunk 8]
```diff
@@ -92,18 +166,29 @@ def lookup_name (address, timeout=10):
...
+
+    Note
+    ----
     Since this affects all Bluetooth connections to that device, and not just
     those initiated by this process or PyBluez, a call to this method requires
     superuser privileges.
 
     You must have an active connection to the specified device before invoking
-    this method.
+    this method
```
Comment (rgov) [inline bluetooth/bluez.py:191]: Are we supposed to remove this period in the new comment style? I don't see it consistently done here. I would prefer the period.

File: bluetooth/bluez.py [hunk 9]
```diff
@@ -143,7 +233,7 @@ def set_l2cap_mtu (sock, mtu):
     that all L2CAP connections start with is 672 bytes.
 
     mtu must be between 48 and 65535, inclusive.
-
+
```
Comment (rgov) [inline bluetooth/bluez.py:236]: Please remove whitespace at the ends of lines. Something like `sed -i '' 's/ *$//' file.py` will do it, I think.

File: bluetooth/bluez.py [hunk 10]
```diff
@@ -158,32 +248,85 @@ def _get_available_ports(protocol):
         return [0]
 
 class BluetoothSocket:
-    __doc__ = _bt.btsocket.__doc__
-
+    """Defines a BluetoothSocket.
+    
+    Instantiating an object of this class open a socket of the given protocol. 
+    The protocol must be one of : HCI, L2CAP, RFCOMM, or SCO. SCO sockets have not been
```
Comment (rgov) [inline bluetooth/bluez.py:254]: Please remove the colon.

File: bluetooth/bluez.py [hunk 11]
```diff
@@ -158,32 +248,85 @@ def _get_available_ports(protocol):
...
+    
+    """
     def __init__ (self, proto = RFCOMM, _sock=None):
+        """Initialise a BluetoothSocket instance.
+        
+        Parameters
+        ----------
+        proto : int, optional
+            The protocol the socket will use. The options are HCI, L2CAP, RFCOMM, or SCO. 
+            The default is RFCOMM
```
Comment (rgov) [inline bluetooth/bluez.py:267]: Missing a period here.

**Review by rgov — CHANGES_REQUESTED** [review]
Sorry for the nitpicks!

**Comment by rgov** [comment]
I'm still concerned about the doc duplication between `__init__.py` and `bluez.py`. It will be easy for these to diverge and then some contributor down the line will have to figure out which one is correct and might get it wrong. I will ask around to other Python communities how to handle this properly.

https://stackoverflow.com/questions/57274925/documenting-a-module-interface-with-multiple-implementations

File: bluetooth/bluez.py (hunk 8 above)
Comment (beadysea) [reply bluetooth/bluez.py:191]: my mistake.

File: bluetooth/bluez.py (hunk 6 above)
Comment (beadysea) [reply bluetooth/bluez.py:14]: You're correct. Not sure what has happened here.

**Comment by beadysea** [comment]
I see your point about duplication. I'm going to have a rethink about how best to go about this.

**Comment by beadysea** [comment]
I've closed this because, on reflection, I think it's the wrong approach to take. It would be much simpler to focus on documenting the API in `bluetooth.__init__.py`, providing a single authoritative reference.

**Comment by rgov** [comment]
Alright. I hope you don't lose too much work and can reuse what you've done
to clean up the docstrings.

What do you think about making __init__.py declare an interface, with each
function just passing arguments through to the backend module? See my
StackOverflow link above, the second example.

**Comment by beadysea** [comment]
My worry with that method is that if someone extends functionality in a backend module, wouldn't they need to remember to add it to__init__.py? On the other hand, I've been experimenting with adding `__doc__` elements to init and sphinx throw errors with some functions. I'm still trying to figure out why but suspect it's to do with sphinx not finding the function's definition. This may turn out to be a good side effect but I need to investigate further.

**Comment by rgov** [comment]
Well, my feeling is that we would probably *want* this to be the case, having to update the interface in `__init__.py`. Then all the other backends need to be updated to specify that the functionality is not implemented.

**Comment by beadysea** [comment]
I think that's why sphinx is throwing errors with the `__doc__` method. for example functions like `get_l2cap_options` aren't defined in msbt.py. There are also functions defined in msbt.py and macos.py that aren't defined in bluez.py. Thinking out loud I'm wondering about how we handle the BluetoothSocket and DeviceDiscoverer classes. I don't think we would want to define them in init.

**Comment by rgov** [comment]
We should define them in those backends and have them raise NotImplementedError. The interface should be the same across platforms as much as we can.

**Comment by ukBaz** [comment]
@rgov, to suggest an answer for your question you posted on StackOverflow, is it a Factory design pattern you are looking for?
More information at:
https://python-3-patterns-idioms-test.readthedocs.io/en/latest/Factory.html

Another suggestion is to look at the [gpiozero project](https://github.com/gpiozero/gpiozero). I always think it is a good reference as to what Pythonic documentation should look like.

**Comment by beadysea** [comment]
Also here: https://realpython.com/factory-method-python/

**Comment by rgov** [comment]
I would say it's related to the factory pattern but that is usually in
reference to object-oriented programming, and I am not sure that I see the
value in converting PyBluez to be more object-oriented at this point.

Think of the C standard library functions, like strcpy(). There is one
standard interface for how these functions work, but depending on your
operating system, those function implementations may be provided by glib
(Linux), libSystem (macOS), msvcrt (Windows), etc.

**Comment by beadysea** [comment]
I'm still not happy with this. I think it's too messy so I'm going to abandon this PR and do things in smaller stages.

**Pull request closed without merging**